pkg/controller/certificates/util_test.go

```diff
@@ -282,67 +283,82 @@ func selfSignCertificate(t *testing.T, spec cmapi.CertificateSpec) []byte {
-			specRenewBefore:                  nil,
-			defaultRenewBeforeExpiryDuration: time.Hour * 24 * 7,
-			expected:                         time.Hour * 8,
+		"no renewBeforeHint, defaultRenewBefore > (cert duration / 3)": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour * 24),
+			renewBeforeHint:     nil,
+			defaultRenewBefore:  time.Hour * 24 * 7,
+			expectedRenewalTime: &metav1.Time{Time: now.Add(time.Hour * 16)},
 		},
-		"spec larger than actual duration": {
-			notBefore:                        now,
-			notAfter:                         now.Add(time.Hour * 24),
-			specRenewBefore:                  &metav1.Duration{Duration: time.Hour * 24 * 7},
-			defaultRenewBeforeExpiryDuration: time.Hour,
-			expected:                         time.Hour * 8,
+		"renewBeforeHint > (cert duration / 3)": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour * 24),
+			renewBeforeHint:     &metav1.Duration{Duration: time.Hour * 24 * 7},
+			defaultRenewBefore:  time.Hour,
+			expectedRenewalTime: &metav1.Time{Time: now.Add(time.Hour * 16)},
 		},
-		"default equal to actual duration": {
-			notBefore:                        now,
-			notAfter:                         now.Add(time.Hour * 24),
-			specRenewBefore:                  nil,
-			defaultRenewBeforeExpiryDuration: time.Hour * 24,
-			expected:                         time.Hour * 8,
+		"no renewBeforeHint, defaultRenewBefore == cert duration": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour * 24),
+			renewBeforeHint:     nil,
+			defaultRenewBefore:  time.Hour * 24,
+			expectedRenewalTime: &metav1.Time{Time: now.Add(time.Hour * 16)},
 		},
-		"spec equal to actual duration": {
-			notBefore:                        now,
-			notAfter:                         now.Add(time.Hour * 24),
-			specRenewBefore:                  &metav1.Duration{Duration: time.Hour * 24},
-			defaultRenewBeforeExpiryDuration: time.Hour,
-			expected:                         time.Hour * 8,
+		"renewBeforeHint == cert duration": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour * 24),
+			renewBeforeHint:     &metav1.Duration{Duration: time.Hour * 24},
+			defaultRenewBefore:  time.Hour,
+			expectedRenewalTime: &metav1.Time{Time: now.Add(time.Hour * 16)},
+		},
+
+		// The following two test cases would catch the bug reported in
+		// https://github.com/jetstack/cert-manager/issues/3897
+		"cert duration very slightly more than defaultRenewBefore": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour*24 + time.Minute*3),
+			renewBeforeHint:     nil,
+			defaultRenewBefore:  time.Hour * 24,
+			expectedRenewalTime: &metav1.Time{Time: now.Add(time.Hour*16 + time.Minute*2)},
+		},
+		"cert duration very slightly more than renewBeforeHint": {
+			notBefore:           now,
+			notAfter:            now.Add(time.Hour*24 + time.Minute*3),
+			renewBeforeHint:     nil,
+			defaultRenewBefore:  time.Hour * 24,
```

Comment:
This case says that it is using `renewBeforeHint`, but `renewBeforeHint` is `nil`. I think this case is identical to the previous one as written, but should be testing the hint case rather than the default.